Fix: Handle no-data values in TIF files during prediction

henrykironde opened the pull request:
Previously, when reading TIF files with rasterio, no-data values were not masked before normalization. This caused invalid values (e.g., -9999 normalized to -39.2) to be passed to the model, potentially affecting prediction accuracy.

## Description

<!-- Provide a clear and concise description of what this PR does -->

## Related Issue(s)

<!-- Link to related issues using: Closes #123, Fixes #456, Related to #789 -->
<!-- If no issue exists, explain why this change is needed -->


## AI-Assisted Development

<!-- Be transparent about AI tool usage -->

- [x] I used AI tools (e.g., GitHub Copilot, ChatGPT, etc.) in developing this PR
- [x] I understand all the code I'm submitting
- [x] I have reviewed and validated all AI-generated code

**AI tools used (if applicable):**
<!-- List AI tools used, if any -->

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Implements consistent no-data handling across raster reads and tightens window bounds to prevent out-of-range access.
> 
> - Adds `utilities.apply_nodata_mask` to read raster windows, clip to bounds, and set no-data pixels to 0 using `dataset_mask`
> - Replaces direct `src.read(...)` with `apply_nodata_mask` in `datasets/cropmodel.py`, `datasets/prediction.py` (TiledRaster), `model.py` (crop writing), and `utilities.crop_raster`
> - Clamps crop windows to image bounds in `BoundingBoxDataset.__getitem__` to avoid out-of-bounds errors
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 8df58ebcbec74c1d1dfa2c384fe9c4c479977513. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

henrykironde committed:
Fix: Handle no-data values in TIF files during prediction

Previously, when reading TIF files with rasterio, no-data values were
not masked before normalization. This caused invalid values (e.g., -9999
normalized to -39.2) to be passed to the model, potentially affecting
prediction accuracy.

ethanwhite reviewed (approved):
LGTM. Nice work @henrykironde!

I'll leave it to either @bw4sz or @jveitchmichaelis to give this a 2nd read and then merge.

henrykironde committed:
Refactor nodata masking to use rasterio's dataset_mask API

- Rename read_raster_window to apply_nodata_mask
- Combine src.read() with masking in single function
- Use rasterio's dataset_mask() for efficient masking
- Update all call sites to new signature

henrykironde committed:
Avoid out-of-bounds errors

henrykironde committed:
rephrasing apply_nodata_mask docsting

jveitchmichaelis commented on src/deepforest/datasets/cropmodel.py:
```diff
@@ -121,7 +123,10 @@ def __getitem__(self, idx):
         width = int(max(1, xmax - xmin))
         height = int(max(1, ymax - ymin))
-        box = self.src.read(window=Window(col_off, row_off, width, height))
+        # Clip window to image bounds to avoid out-of-bounds errors
```
Test case for this would be useful - add a dataframe with boxes that extend outside the image and confirm this doesn't break.

jveitchmichaelis reviewed (changes requested):
Thanks @henrykironde it looks good. I'd request a couple of unit tests to check some of the bounds checking logic. I don't think we need @bw4sz 's failing test first, just some sample nodata.
